httpcore5/src/main/java/org/apache/hc/core5/pool/StrictConnPool.java

```diff
@@ -172,7 +175,20 @@ public void close() {
         Args.notNull(requestTimeout, "Request timeout");
         Asserts.check(!this.isShutDown.get(), "Connection pool shut down");
         final Deadline deadline = Deadline.calculate(requestTimeout);
-        final BasicFuture<PoolEntry<T, C>> future = new BasicFuture<>(callback);
+        final BasicFuture<PoolEntry<T, C>> future = new BasicFuture<PoolEntry<T, C>>(callback) {
+
+            @Override
+            public synchronized PoolEntry<T, C> get(
+                    final long timeout, final TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+                try {
+                    return super.get(timeout, unit);
+                } catch (final TimeoutException ex) {
+                    cancel();
+                    throw ex;
+                }
+            }
+
+        };
```

Comment:
You'd have less boilerplate using lambdas I believe. On my phone here but the code seems duplicate. Can it be refactored or even built as a singleton?

Reply:
Bad abstractions are a bigger maintenance burden than duplication.